src/emulator/storage/upload.ts

```diff
@@ -0,0 +1,188 @@
+  bucketId: string;
+  objectId: string;
+  metadataRaw: string;
+  contentType: string;
+  authorization?: string;
+};
+
+/** Error that signals a resumable upload that's expected to be active is not. */
+export class UploadNotActiveError extends Error {}
+
+/** Error that signals a resumable upload is not cancellable.  */
+export class NotCancellableError extends Error {}
+
+/**
+ * Service that handles byte transfer and maintains state for file uploads.
+ *
+ * New file uploads will be persisted to a temp staging directory which will not
+ * survive across emulator restarts. Clients are expected to move staged files
+ * to a more permanent location.
+ */
+export class UploadService {
+  private _uploads!: Map<string, Upload>;
+  constructor(private _persistence: Persistence) {
+    this.reset();
+  }
+
+  /** Resets the state of the UploadService. */
+  public reset(): void {
+    this._uploads = new Map();
+  }
+
+  /**
+   * Handles a multipart file upload which is expected to have the entirety of
+   * the file's contents in a single request.
+   */
+  public multipartUpload(request: MultipartUploadRequest): Upload {
+    const data = Buffer.from(request.dataRaw);
+    const upload = this.initMultipartUpload(request, data.byteLength);
+    this._persistence.deleteFile(upload.path, /* failSilently = */ true);
+    this._persistence.appendBytes(upload.path, data);
+    return upload;
+  }
+
+  private initMultipartUpload(request: MultipartUploadRequest, sizeInBytes: number): Upload {
+    const id = uuidV4();
+    const upload: Upload = {
+      id: uuidV4(),
+      bucketId: request.bucketId,
+      objectId: request.objectId,
+      type: UploadType.MULTIPART,
+      path: this.stagingFileName(id, request.bucketId, request.objectId),
+      status: UploadStatus.FINISHED,
+      metadata: JSON.parse(request.metadataRaw),
+      size: sizeInBytes,
+      authorization: request.authorization,
+    };
+    this._uploads.set(upload.id, upload);
+
+    return upload;
+  }
+
+  /**
+   * Initializes a new ResumableUpload.
+   */
+  public startResumableUpload(request: StartResumableUploadRequest): Upload {
+    const id = uuidV4();
+    const upload: Upload = {
+      id: id,
+      bucketId: request.bucketId,
+      objectId: request.objectId,
+      type: UploadType.RESUMABLE,
+      path: this.stagingFileName(id, request.bucketId, request.objectId),
+      status: UploadStatus.ACTIVE,
+      metadata: JSON.parse(request.metadataRaw),
+      size: 0,
+      authorization: request.authorization,
+    };
+    this._uploads.set(upload.id, upload);
+    this._persistence.deleteFile(upload.path, /* failSilently = */ true);
+    return upload;
+  }
+
+  /**
+   * Appends bytes to an existing resumable upload.
+   * @throws {NotFoundError} if the resumable upload does not exist.
+   * @throws {NotActiveUploadError} if the resumable upload is not in the ACTIVE state.
+   */
+  public progressResumableUpload(uploadId: string, dataRaw: string): Upload {
```

Comment:
I don't think progress is a transitive verb 😅 What about continue?

Reply:
done